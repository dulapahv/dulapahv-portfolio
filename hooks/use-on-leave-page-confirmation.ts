'use client';

import { useEffect, useRef } from 'react';
import { usePathname, useRouter } from 'next/navigation';

const MESSAGE = 'Changes you made may not be saved.';

export function useOnLeavePageConfirmation(shouldConfirm: boolean) {
  const pathname = usePathname();
  const router = useRouter();

  const isInitialMount = useRef(true);

  // Handle refresh or close tab/window
  useEffect(() => {
    if (!shouldConfirm) return;

    function beforeUnload(e: BeforeUnloadEvent) {
      e.preventDefault();
      e.returnValue = '';
    }

    window.addEventListener('beforeunload', beforeUnload);
    return () => window.removeEventListener('beforeunload', beforeUnload);
  }, [shouldConfirm]);

  // Handle back/forward browser buttons
  useEffect(() => {
    if (!shouldConfirm) return;

    const handleBackButton = () => {
      if (!window.confirm(MESSAGE)) history.pushState(null, '', window.location.href);
      else {
        window.removeEventListener('popstate', handleBackButton);
        history.back();
        // When user clicks back after refreshing the page with unsaved changes
        setTimeout(() => {
          if (window.location.pathname === pathname) router.back();
        }, 0);
      }
    };

    // Push initial state to history to detect back navigation
    if (isInitialMount.current) {
      history.pushState(null, '', window.location.href);
      isInitialMount.current = false;
    }

    window.addEventListener('popstate', handleBackButton);
    return () => window.removeEventListener('popstate', handleBackButton);
  }, [shouldConfirm, pathname]);

  // Handle Next.js Link navigation
  useEffect(() => {
    if (!shouldConfirm) return;

    const handleClick = (e: MouseEvent) => {
      const target = e.target as HTMLElement;
      const anchor = target.closest('a');

      if (anchor && anchor.href && anchor.href.startsWith(window.location.origin)) {
        const targetUrl = new URL(anchor.href);
        const currentPath = pathname;

        if (targetUrl.pathname !== currentPath) {
          if (!window.confirm(MESSAGE)) {
            e.preventDefault();
            e.stopPropagation();
          }
        }
      }
    };

    document.addEventListener('click', handleClick, true);
    return () => document.removeEventListener('click', handleClick, true);
  }, [shouldConfirm, pathname]);
}
