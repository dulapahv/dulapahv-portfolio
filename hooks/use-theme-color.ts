"use client";

import { useEffect } from 'react';

import { useTheme } from 'next-themes';

export function useThemeColor() {
  const { resolvedTheme } = useTheme();

  useEffect(() => {
    // Get the theme color meta tag
    const themeColorMeta = document.querySelector('meta[name="theme-color"]');

    if (themeColorMeta) {
      // Update the color based on theme
      themeColorMeta.setAttribute(
        'content',
        resolvedTheme === 'dark' ? '#0a0a0a' : '#eef1f7',
      );
    }
  }, [resolvedTheme]);
}
