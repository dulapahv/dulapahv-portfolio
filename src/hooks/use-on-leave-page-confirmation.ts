"use client";

import { useEffect } from "react";

export function useOnLeavePageConfirmation(isDirty: boolean) {
  useEffect(() => {
    function beforeUnload(e: BeforeUnloadEvent) {
      if (!isDirty) return;
      e.preventDefault();
    }

    window.addEventListener("beforeunload", beforeUnload);

    return () => {
      window.removeEventListener("beforeunload", beforeUnload);
    };
  }, [isDirty]);
}
