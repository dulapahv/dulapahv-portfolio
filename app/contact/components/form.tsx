'use client';

import { useRef, useState, type FormEventHandler } from 'react';

import { Turnstile, type TurnstileInstance } from '@marsidev/react-turnstile';
import { PaperPlaneTiltIcon } from '@phosphor-icons/react/dist/ssr';
import { Form } from 'radix-ui';

import {
  EMAIL_MAX_LENGTH,
  EMAIL_REGEX,
  MESSAGE_MAX_LENGTH,
  NAME_MAX_LENGTH,
  TURNSTILE_SITE_KEY,
} from '@/lib/constants';
import { cn } from '@/lib/utils';
import { useOnLeavePageConfirmation } from '@/hooks/use-on-leave-page-confirmation';
import type { RecipientEmailTemplateProps } from '@/components/email';
import { Input } from '@/components/input';
import { Spinner } from '@/components/spinner';
import { Textarea } from '@/components/textarea';
import { sendContactEmail } from '@/app/actions/contact';

interface ContactFormProps {
  searchParams?: RecipientEmailTemplateProps;
}

export const ContactForm = ({ searchParams }: ContactFormProps) => {
  const initialName = searchParams?.name || '';
  const initialEmail = searchParams?.email || '';
  const initialMessage = searchParams?.message || '';

  const [isPending, setIsPending] = useState(false);
  const [submitError, setSubmitError] = useState<string>('');
  const [submitSuccess, setSubmitSuccess] = useState(false);
  const [isTurnstileSolved, setIsTurnstileSolved] = useState(false);
  const [isFormDirty, setIsFormDirty] = useState(false);

  const formRef = useRef<HTMLFormElement>(null);
  const turnstileRef = useRef<TurnstileInstance>(null);

  const handleSubmit: FormEventHandler<HTMLFormElement> = async (event) => {
    event.preventDefault();
    if (!isTurnstileSolved || isPending) {
      setSubmitError('Please complete the CAPTCHA to verify you are human.');
      return;
    }
    setSubmitError('');
    setSubmitSuccess(false);
    const formData = new FormData(event.currentTarget);
    const data = {
      name: formData.get('name') as string,
      email: formData.get('email') as string,
      message: formData.get('message') as string,
    };
    try {
      setIsPending(true);
      const result = await sendContactEmail(data);
      if (result.error) {
        setSubmitError(result.error);
        return;
      }
      setSubmitSuccess(true);
      formRef.current?.reset();
    } catch (error) {
      setSubmitError(
        `An unexpected error occurred. Please try again.\n${error}`,
      );
    } finally {
      setIsPending(false);
      setIsTurnstileSolved(false);
      turnstileRef.current?.reset();
    }
  };

  useOnLeavePageConfirmation(isFormDirty);

  return (
    <Form.Root
      ref={formRef}
      onSubmit={handleSubmit}
      onChange={() => setIsFormDirty(true)}
      className="grid w-full gap-6"
      role="form"
      aria-label="Contact form"
    >
      <Input
        label="Name"
        name="name"
        autoComplete="name"
        placeholder="Dulapah Vibulsanti"
        required
        maxLength={NAME_MAX_LENGTH}
        defaultValue={initialName}
      />
      <Input
        label="Email"
        name="email"
        type="email"
        autoComplete="email"
        placeholder="dulapah@example.com"
        pattern={EMAIL_REGEX.source}
        required
        maxLength={EMAIL_MAX_LENGTH}
        defaultValue={initialEmail}
      />
      <Textarea
        label="Message"
        name="message"
        placeholder="Hi Dulapah, I wanted to reach out to you about..."
        required
        maxLength={MESSAGE_MAX_LENGTH}
        defaultValue={initialMessage}
      />

      {/* Success/Error Messages */}
      {submitSuccess && (
        <div
          className="text-success text-sm font-semibold"
          role="status"
          aria-live="polite"
        >
          Your message has been sent successfully! You will receive a
          confirmation email and hear back from me soon.
        </div>
      )}
      {submitError && (
        <div
          className="text-error text-sm font-semibold whitespace-pre-line"
          role="alert"
          aria-live="assertive"
        >
          {submitError}
        </div>
      )}

      <div className="grid gap-1">
        <span
          id="turnstile-label"
          className="text-foreground-muted inline-block text-sm font-medium"
        >
          Let me know you&apos;re human
          <span className="after:text-error after:ml-0.5 after:content-['*']">
            <span className="sr-only"> (required)</span>
          </span>
        </span>
        <div role="group" aria-labelledby="turnstile-label">
          <Turnstile
            ref={turnstileRef}
            id="turnstile-widget"
            siteKey={TURNSTILE_SITE_KEY}
            onSuccess={() => setIsTurnstileSolved(true)}
            onExpire={() => {
              setIsTurnstileSolved(false);
              turnstileRef.current?.reset();
            }}
          />
        </div>
      </div>

      <Form.Submit asChild>
        <button
          type="submit"
          disabled={!isTurnstileSolved || isPending}
          aria-disabled={!isTurnstileSolved || isPending}
          aria-describedby={
            submitError
              ? 'submit-error'
              : submitSuccess
                ? 'submit-success'
                : undefined
          }
          className={cn(
            `bg-mirai-red flex w-fit cursor-pointer items-center justify-center gap-2
            rounded-md px-3 py-2 text-sm font-medium text-white transition-transform
            select-none`,
            'hover:bg-mirai-red/90 hover:shadow-lg',
            'active:scale-[0.98] active:shadow-md',
            'disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none',
          )}
        >
          <>
            {isPending ? (
              <Spinner className="relative" />
            ) : (
              <PaperPlaneTiltIcon className="size-4.5" aria-hidden="true" />
            )}
            <span>{isPending ? 'Sending...' : 'Send Message'}</span>
          </>
        </button>
      </Form.Submit>
    </Form.Root>
  );
};
