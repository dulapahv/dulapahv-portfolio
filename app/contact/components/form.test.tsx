import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { useOnLeavePageConfirmation } from '@/hooks/use-on-leave-page-confirmation';

import { ContactForm } from './form';

// Mock dependencies
vi.mock('@/hooks/use-on-leave-page-confirmation', () => ({
  useOnLeavePageConfirmation: vi.fn()
}));

vi.mock('@/app/actions/contact', () => ({
  sendContactEmail: vi.fn()
}));

// Mock Turnstile
vi.mock('@marsidev/react-turnstile', () => ({
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  Turnstile: ({ onSuccess }: any) => (
    <div
      data-testid="turnstile-mock"
      onClick={() => onSuccess && onSuccess()}
      role="button"
      aria-label="CAPTCHA verification"
    >
      Turnstile Mock
    </div>
  )
}));

// Mock radix-ui Form
vi.mock('radix-ui', () => ({
  Form: {
    /* eslint-disable  @typescript-eslint/no-explicit-any */
    Root: ({ children, ...props }: any) => <form {...props}>{children}</form>,
    /* eslint-disable  @typescript-eslint/no-explicit-any */
    Submit: ({ children }: any) => <>{children}</>
  }
}));

// Mock components
vi.mock('@/components/input', () => ({
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  Input: ({ label, ...props }: any) => (
    <div>
      <label htmlFor={props.name}>{label}</label>
      <input id={props.name} {...props} />
    </div>
  )
}));

vi.mock('@/components/textarea', () => ({
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  Textarea: ({ label, ...props }: any) => (
    <div>
      <label htmlFor={props.name}>{label}</label>
      <textarea id={props.name} {...props} />
    </div>
  )
}));

vi.mock('@/components/spinner', () => ({
  /* eslint-disable  @typescript-eslint/no-explicit-any */
  Spinner: ({ ...props }: any) => <span {...props}>Loading...</span>
}));

describe('ContactForm', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('Rendering', () => {
    it('should render the contact form', () => {
      render(<ContactForm />);

      expect(screen.getByRole('form', { name: /contact form/i })).toBeInTheDocument();
    });

    it('should render name input field', () => {
      render(<ContactForm />);

      expect(screen.getByLabelText(/name/i)).toBeInTheDocument();
    });

    it('should render email input field', () => {
      render(<ContactForm />);

      expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
    });

    it('should render message textarea field', () => {
      render(<ContactForm />);

      expect(screen.getByLabelText(/message/i)).toBeInTheDocument();
    });

    it('should render Turnstile CAPTCHA', () => {
      render(<ContactForm />);

      expect(screen.getByTestId('turnstile-mock')).toBeInTheDocument();
    });

    it('should render submit button', () => {
      render(<ContactForm />);

      expect(screen.getByRole('button', { name: /send message/i })).toBeInTheDocument();
    });

    it('should have submit button disabled initially', () => {
      render(<ContactForm />);

      const submitButton = screen.getByRole('button', { name: /send message/i });
      expect(submitButton).toBeDisabled();
    });
  });

  describe('Form fields with initial values', () => {
    it('should populate name field from searchParams', () => {
      const searchParams = {
        name: 'John Doe',
        email: '',
        message: ''
      };

      render(<ContactForm searchParams={searchParams} />);

      const nameInput = screen.getByLabelText(/name/i) as HTMLInputElement;
      expect(nameInput.value).toBe('John Doe');
    });

    it('should populate email field from searchParams', () => {
      const searchParams = {
        name: '',
        email: 'john@example.com',
        message: ''
      };

      render(<ContactForm searchParams={searchParams} />);

      const emailInput = screen.getByLabelText(/email/i) as HTMLInputElement;
      expect(emailInput.value).toBe('john@example.com');
    });

    it('should populate message field from searchParams', () => {
      const searchParams = {
        name: '',
        email: '',
        message: 'Hello, this is a test message'
      };

      render(<ContactForm searchParams={searchParams} />);

      const messageInput = screen.getByLabelText(/message/i) as HTMLTextAreaElement;
      expect(messageInput.value).toBe('Hello, this is a test message');
    });

    it('should populate all fields from searchParams', () => {
      const searchParams = {
        name: 'Dulapah',
        email: 'dulapah@example.com',
        message: 'Test message'
      };

      render(<ContactForm searchParams={searchParams} />);

      const nameInput = screen.getByLabelText(/name/i) as HTMLInputElement;
      const emailInput = screen.getByLabelText(/email/i) as HTMLInputElement;
      const messageInput = screen.getByLabelText(/message/i) as HTMLTextAreaElement;

      expect(nameInput.value).toBe('Dulapah');
      expect(emailInput.value).toBe('dulapah@example.com');
      expect(messageInput.value).toBe('Test message');
    });
  });

  describe('CAPTCHA interaction', () => {
    it('should enable submit button after solving CAPTCHA', async () => {
      render(<ContactForm />);

      const submitButton = screen.getByRole('button', { name: /send message/i });
      expect(submitButton).toBeDisabled();

      // Solve CAPTCHA
      const turnstile = screen.getByTestId('turnstile-mock');
      fireEvent.click(turnstile);

      await waitFor(() => {
        expect(submitButton).not.toBeDisabled();
      });
    });
  });

  describe('Form submission', () => {
    beforeEach(async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      vi.mocked(sendContactEmail).mockResolvedValue({ success: true });
    });

    it('should show error if submitting without solving CAPTCHA', async () => {
      render(<ContactForm />);

      const form = screen.getByRole('form');
      fireEvent.submit(form);

      await waitFor(() => {
        expect(screen.getByRole('alert')).toBeInTheDocument();
        expect(screen.getByRole('alert').textContent).toContain('Please complete the CAPTCHA');
      });
    });

    it('should submit form successfully when CAPTCHA is solved', async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      vi.mocked(sendContactEmail).mockResolvedValue({ success: true });

      render(<ContactForm />);

      // Fill form
      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John Doe' } });
      fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'john@example.com' } });
      fireEvent.change(screen.getByLabelText(/message/i), { target: { value: 'Test message' } });

      // Solve CAPTCHA
      fireEvent.click(screen.getByTestId('turnstile-mock'));

      // Submit form
      await waitFor(() => {
        const submitButton = screen.getByRole('button', { name: /send message/i });
        expect(submitButton).not.toBeDisabled();
      });

      const form = screen.getByRole('form');
      fireEvent.submit(form);

      await waitFor(() => {
        expect(sendContactEmail).toHaveBeenCalledWith({
          name: 'John Doe',
          email: 'john@example.com',
          message: 'Test message'
        });
      });
    });

    it('should show success message after successful submission', async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      vi.mocked(sendContactEmail).mockResolvedValue({ success: true });

      render(<ContactForm />);

      // Fill and submit form
      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John Doe' } });
      fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'john@example.com' } });
      fireEvent.change(screen.getByLabelText(/message/i), { target: { value: 'Test message' } });
      fireEvent.click(screen.getByTestId('turnstile-mock'));

      await waitFor(() => {
        fireEvent.submit(screen.getByRole('form'));
      });

      await waitFor(() => {
        expect(screen.getByRole('status')).toBeInTheDocument();
        expect(screen.getByRole('status').textContent).toContain(
          'Your message has been sent successfully'
        );
      });
    });

    it('should show error message on submission failure', async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      vi.mocked(sendContactEmail).mockResolvedValue({ error: 'Failed to send email' });

      render(<ContactForm />);

      // Fill and submit form
      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John Doe' } });
      fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'john@example.com' } });
      fireEvent.change(screen.getByLabelText(/message/i), { target: { value: 'Test message' } });
      fireEvent.click(screen.getByTestId('turnstile-mock'));

      await waitFor(() => {
        fireEvent.submit(screen.getByRole('form'));
      });

      await waitFor(() => {
        expect(screen.getByRole('alert')).toBeInTheDocument();
        expect(screen.getByRole('alert').textContent).toContain('Failed to send email');
      });
    });

    it('should show error message on unexpected error', async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      vi.mocked(sendContactEmail).mockRejectedValue(new Error('Network error'));

      render(<ContactForm />);

      // Fill and submit form
      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John Doe' } });
      fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'john@example.com' } });
      fireEvent.change(screen.getByLabelText(/message/i), { target: { value: 'Test message' } });
      fireEvent.click(screen.getByTestId('turnstile-mock'));

      await waitFor(() => {
        fireEvent.submit(screen.getByRole('form'));
      });

      await waitFor(() => {
        expect(screen.getByRole('alert')).toBeInTheDocument();
        expect(screen.getByRole('alert').textContent).toContain('An unexpected error occurred');
      });
    });
  });

  describe('Loading state', () => {
    it('should show loading text during submission', async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      /* eslint-disable  @typescript-eslint/no-explicit-any */
      let resolveSubmit: any;
      vi.mocked(sendContactEmail).mockReturnValue(
        new Promise(resolve => {
          resolveSubmit = resolve;
        })
      );

      render(<ContactForm />);

      // Fill and submit form
      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John Doe' } });
      fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'john@example.com' } });
      fireEvent.change(screen.getByLabelText(/message/i), { target: { value: 'Test message' } });
      fireEvent.click(screen.getByTestId('turnstile-mock'));

      await waitFor(() => {
        fireEvent.submit(screen.getByRole('form'));
      });

      await waitFor(() => {
        expect(screen.getByRole('button', { name: /sending/i })).toBeInTheDocument();
      });

      // Resolve the submission
      resolveSubmit({ success: true });
    });

    it('should disable submit button during submission', async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      /* eslint-disable  @typescript-eslint/no-explicit-any */
      let resolveSubmit: any;
      vi.mocked(sendContactEmail).mockReturnValue(
        new Promise(resolve => {
          resolveSubmit = resolve;
        })
      );

      render(<ContactForm />);

      // Fill and submit form
      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John Doe' } });
      fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'john@example.com' } });
      fireEvent.change(screen.getByLabelText(/message/i), { target: { value: 'Test message' } });
      fireEvent.click(screen.getByTestId('turnstile-mock'));

      await waitFor(() => {
        fireEvent.submit(screen.getByRole('form'));
      });

      await waitFor(() => {
        const button = screen.getByRole('button', { name: /sending/i });
        expect(button).toBeDisabled();
      });

      // Resolve the submission
      resolveSubmit({ success: true });
    });
  });

  describe('Form dirty state', () => {
    it('should call useOnLeavePageConfirmation hook with false initially', () => {
      render(<ContactForm />);

      expect(useOnLeavePageConfirmation).toHaveBeenCalledWith(false);
    });

    it('should call useOnLeavePageConfirmation hook with true after form change', async () => {
      render(<ContactForm />);

      // Change a form field
      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John' } });

      await waitFor(() => {
        expect(useOnLeavePageConfirmation).toHaveBeenCalledWith(true);
      });
    });
  });

  describe('Accessibility', () => {
    it('should have accessible form label', () => {
      render(<ContactForm />);

      expect(screen.getByRole('form')).toHaveAttribute('aria-label', 'Contact form');
    });

    it('should have required indicator for name field', () => {
      render(<ContactForm />);

      const nameInput = screen.getByLabelText(/name/i);
      expect(nameInput).toBeRequired();
    });

    it('should have required indicator for email field', () => {
      render(<ContactForm />);

      const emailInput = screen.getByLabelText(/email/i);
      expect(emailInput).toBeRequired();
    });

    it('should have required indicator for message field', () => {
      render(<ContactForm />);

      const messageInput = screen.getByLabelText(/message/i);
      expect(messageInput).toBeRequired();
    });

    it('should have aria-live for success message', async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      vi.mocked(sendContactEmail).mockResolvedValue({ success: true });

      render(<ContactForm />);

      // Fill and submit form
      fireEvent.change(screen.getByLabelText(/name/i), { target: { value: 'John Doe' } });
      fireEvent.change(screen.getByLabelText(/email/i), { target: { value: 'john@example.com' } });
      fireEvent.change(screen.getByLabelText(/message/i), { target: { value: 'Test message' } });
      fireEvent.click(screen.getByTestId('turnstile-mock'));

      await waitFor(() => {
        fireEvent.submit(screen.getByRole('form'));
      });

      await waitFor(() => {
        const successMessage = screen.getByRole('status');
        expect(successMessage).toHaveAttribute('aria-live', 'polite');
      });
    });

    it('should have aria-live for error message', async () => {
      render(<ContactForm />);

      // Submit without CAPTCHA
      fireEvent.submit(screen.getByRole('form'));

      await waitFor(() => {
        const errorMessage = screen.getByRole('alert');
        expect(errorMessage).toHaveAttribute('aria-live', 'assertive');
      });
    });
  });

  describe('Form reset', () => {
    it('should reset form after successful submission', async () => {
      const { sendContactEmail } = await import('@/app/actions/contact');
      vi.mocked(sendContactEmail).mockResolvedValue({ success: true });

      render(<ContactForm />);

      // Fill form
      const nameInput = screen.getByLabelText(/name/i) as HTMLInputElement;
      const emailInput = screen.getByLabelText(/email/i) as HTMLInputElement;
      const messageInput = screen.getByLabelText(/message/i) as HTMLTextAreaElement;

      fireEvent.change(nameInput, { target: { value: 'John Doe' } });
      fireEvent.change(emailInput, { target: { value: 'john@example.com' } });
      fireEvent.change(messageInput, { target: { value: 'Test message' } });

      // Solve CAPTCHA and submit
      fireEvent.click(screen.getByTestId('turnstile-mock'));

      await waitFor(() => {
        fireEvent.submit(screen.getByRole('form'));
      });

      await waitFor(() => {
        expect(screen.getByRole('status')).toBeInTheDocument();
      });

      // Check that form is reset
      await waitFor(() => {
        expect(nameInput.value).toBe('');
        expect(emailInput.value).toBe('');
        expect(messageInput.value).toBe('');
      });
    });
  });

  describe('Snapshot', () => {
    it('should match snapshot', () => {
      const { container } = render(<ContactForm />);
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with initial values', () => {
      const searchParams = {
        name: 'John Doe',
        email: 'john@example.com',
        message: 'Hello'
      };

      const { container } = render(<ContactForm searchParams={searchParams} />);
      expect(container).toMatchSnapshot();
    });
  });
});
