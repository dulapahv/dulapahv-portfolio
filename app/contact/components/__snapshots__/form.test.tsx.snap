// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContactForm > Snapshot > should match snapshot 1`] = `
<div>
  <form
    aria-label="Contact form"
    class="grid w-full gap-6"
    role="form"
  >
    <div>
      <label
        for="name"
      >
        Name
      </label>
      <input
        autocomplete="name"
        id="name"
        maxlength="256"
        name="name"
        placeholder="Dulapah Vibulsanti"
        required=""
        value=""
      />
    </div>
    <div>
      <label
        for="email"
      >
        Email
      </label>
      <input
        autocomplete="email"
        id="email"
        maxlength="256"
        name="email"
        pattern="^\\S+@\\S+\\.\\S+$"
        placeholder="dulapah@example.com"
        required=""
        type="email"
        value=""
      />
    </div>
    <div>
      <label
        for="message"
      >
        Message
      </label>
      <textarea
        id="message"
        maxlength="2000"
        name="message"
        placeholder="Hi Dulapah, I wanted to reach out to you about..."
        required=""
      />
    </div>
    <div
      class="grid gap-1"
    >
      <span
        class="text-foreground-muted inline-block text-sm font-medium"
        id="turnstile-label"
      >
        Let me know you're human
        <span
          class="after:text-error after:ml-0.5 after:content-['*']"
        >
          <span
            class="sr-only"
          >
             (required)
          </span>
        </span>
      </span>
      <div
        aria-labelledby="turnstile-label"
        role="group"
      >
        <div
          aria-label="CAPTCHA verification"
          data-testid="turnstile-mock"
          role="button"
        >
          Turnstile Mock
        </div>
      </div>
    </div>
    <button
      aria-disabled="true"
      class="bg-mirai-red flex w-fit cursor-pointer items-center justify-center gap-2 rounded-md px-3 py-2 text-sm font-medium text-white transition-transform select-none hover:bg-mirai-red/90 hover:shadow-lg active:scale-[0.98] active:shadow-md disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none"
      disabled=""
      type="submit"
    >
      <svg
        aria-hidden="true"
        class="size-4.5"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M227.32,28.68a16,16,0,0,0-15.66-4.08l-.15,0L19.57,82.84a16,16,0,0,0-2.49,29.8L102,154l41.3,84.87A15.86,15.86,0,0,0,157.74,248q.69,0,1.38-.06a15.88,15.88,0,0,0,14-11.51l58.2-191.94c0-.05,0-.1,0-.15A16,16,0,0,0,227.32,28.68ZM157.83,231.85l-.05.14,0-.07-40.06-82.3,48-48a8,8,0,0,0-11.31-11.31l-48,48L24.08,98.25l-.07,0,.14,0L216,40Z"
        />
      </svg>
      <span>
        Send Message
      </span>
    </button>
  </form>
</div>
`;

exports[`ContactForm > Snapshot > should match snapshot with initial values 1`] = `
<div>
  <form
    aria-label="Contact form"
    class="grid w-full gap-6"
    role="form"
  >
    <div>
      <label
        for="name"
      >
        Name
      </label>
      <input
        autocomplete="name"
        id="name"
        maxlength="256"
        name="name"
        placeholder="Dulapah Vibulsanti"
        required=""
        value="John Doe"
      />
    </div>
    <div>
      <label
        for="email"
      >
        Email
      </label>
      <input
        autocomplete="email"
        id="email"
        maxlength="256"
        name="email"
        pattern="^\\S+@\\S+\\.\\S+$"
        placeholder="dulapah@example.com"
        required=""
        type="email"
        value="john@example.com"
      />
    </div>
    <div>
      <label
        for="message"
      >
        Message
      </label>
      <textarea
        id="message"
        maxlength="2000"
        name="message"
        placeholder="Hi Dulapah, I wanted to reach out to you about..."
        required=""
      >
        Hello
      </textarea>
    </div>
    <div
      class="grid gap-1"
    >
      <span
        class="text-foreground-muted inline-block text-sm font-medium"
        id="turnstile-label"
      >
        Let me know you're human
        <span
          class="after:text-error after:ml-0.5 after:content-['*']"
        >
          <span
            class="sr-only"
          >
             (required)
          </span>
        </span>
      </span>
      <div
        aria-labelledby="turnstile-label"
        role="group"
      >
        <div
          aria-label="CAPTCHA verification"
          data-testid="turnstile-mock"
          role="button"
        >
          Turnstile Mock
        </div>
      </div>
    </div>
    <button
      aria-disabled="true"
      class="bg-mirai-red flex w-fit cursor-pointer items-center justify-center gap-2 rounded-md px-3 py-2 text-sm font-medium text-white transition-transform select-none hover:bg-mirai-red/90 hover:shadow-lg active:scale-[0.98] active:shadow-md disabled:cursor-not-allowed disabled:opacity-50 disabled:hover:shadow-none"
      disabled=""
      type="submit"
    >
      <svg
        aria-hidden="true"
        class="size-4.5"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M227.32,28.68a16,16,0,0,0-15.66-4.08l-.15,0L19.57,82.84a16,16,0,0,0-2.49,29.8L102,154l41.3,84.87A15.86,15.86,0,0,0,157.74,248q.69,0,1.38-.06a15.88,15.88,0,0,0,14-11.51l58.2-191.94c0-.05,0-.1,0-.15A16,16,0,0,0,227.32,28.68ZM157.83,231.85l-.05.14,0-.07-40.06-82.3,48-48a8,8,0,0,0-11.31-11.31l-48,48L24.08,98.25l-.07,0,.14,0L216,40Z"
        />
      </svg>
      <span>
        Send Message
      </span>
    </button>
  </form>
</div>
`;
